tests/test_clean.py

```diff
@@ -1,96 +1,210 @@
-    )
+    with pytest.raises(TypeError) as e:
+        clean(no_type)
+    assert "NoneType" in str(e)
 
 
-def test_with_comments():
-    text = '<!-- comment -->Just text'
-    assert bleach.clean(text) == 'Just text'
-    assert bleach.clean(text, strip_comments=False) == text
+def test_empty():
+    assert clean('') == ''
 
 
-def test_no_html():
-    assert bleach.clean('no html string') == 'no html string'
+def test_content_has_no_html():
+    assert clean('no html string') == 'no html string'
 
 
-def test_allowed_html():
-    assert (
-        bleach.clean('an <strong>allowed</strong> tag') ==
+@pytest.mark.parametrize('data, expected', [
+    (
+        'an <strong>allowed</strong> tag',
         'an <strong>allowed</strong> tag'
-    )
-    assert (
-        bleach.clean('another <em>good</em> tag') ==
+    ),
+
+    (
+        'another <em>good</em> tag',
         'another <em>good</em> tag'
     )
+])
+def test_content_has_allowed_html(data, expected):
+    assert clean(data) == expected
 
 
-def test_bad_html():
+def test_html_is_lowercased():
     assert (
-        bleach.clean('a <em>fixed tag') ==
-        'a <em>fixed tag</em>'
+        clean('<A HREF="http://example.com">foo</A>') ==
+        '<a href="http://example.com">foo</a>'
     )
 
 
-def test_function_arguments():
-    TAGS = ['span', 'br']
-    ATTRS = {'span': ['style']}
+@pytest.mark.parametrize('data, should_strip, expected', [
+    # Regular comment
+    (
+        '<!-- this is a comment -->',
+        True,
+        ''
+    ),
 
-    text = 'a <br/><span style="color:red">test</span>'
-    assert (
-        bleach.clean(text, tags=TAGS, attributes=ATTRS) ==
-        'a <br><span style="">test</span>'
+    # Open comment with no close comment bit
+    (
+        '<!-- open comment',
+        True,
+        ''
+    ),
+    (
+        '<!--open comment',
+        True,
+        ''
+    ),
+    (
+        '<!-- open comment',
+        False,
+        '<!-- open comment-->'
+    ),
+    (
+        '<!--open comment',
+        False,
+        '<!--open comment-->'
+    ),
+
+    # Comment with text to the right
+    (
+        '<!-- comment -->text',
+        True,
+        'text'
+    ),
+    (
+        '<!--comment-->text',
+        True,
+        'text'
+    ),
+    (
+        '<!-- comment -->text',
+        False,
+        '<!-- comment -->text'
+    ),
+    (
+        '<!--comment-->text',
+        False,
+        '<!--comment-->text'
+    ),
+
+    # Comment with text to the left
+    (
+        'text<!-- comment -->',
+        True,
+        'text'
+    ),
+    (
+        'text<!--comment-->',
+        True,
+        'text'
+    ),
+    (
+        'text<!-- comment -->',
+        False,
+        'text<!-- comment -->'
+    ),
+    (
+        'text<!--comment-->',
+        False,
+        'text<!--comment-->'
     )
+])
+def test_comments(data, should_strip, expected):
+    assert clean(data, strip_comments=should_strip) == expected
```

Comment:
We had a bunch of comment tests--this merges them all into one parametrized test. Also, I added some additional test cases here around location of whitespace.

Reply:
👍